packet/conn.go

```diff
@@ -167,6 +144,41 @@ func (c *Conn) ReadPacketReuseMem(dst []byte) ([]byte, error) {
 	return result, nil
 }
 
+// newCompressedPacketReader creates a new compressed packet reader.
+func (c *Conn) newCompressedPacketReader() (io.Reader, error) {
+	if _, err := io.ReadFull(c.reader, c.compressedHeader[:7]); err != nil {
+		return nil, errors.Wrapf(ErrBadConn, "io.ReadFull(compressedHeader) failed. err %v", err)
+	}
+
+	compressedSequence := c.compressedHeader[3]
+	if compressedSequence != c.CompressedSequence {
+		return nil, errors.Errorf("invalid compressed sequence %d != %d",
+			compressedSequence, c.CompressedSequence)
+	}
+
+	compressedLength := int(uint32(c.compressedHeader[0]) | uint32(c.compressedHeader[1])<<8 | uint32(c.compressedHeader[2])<<16)
+	uncompressedLength := int(uint32(c.compressedHeader[4]) | uint32(c.compressedHeader[5])<<8 | uint32(c.compressedHeader[6])<<16)
+	if uncompressedLength > 0 {
+		limitedReader := io.LimitReader(c.reader, int64(compressedLength))
+		switch c.Compression {
+		case MYSQL_COMPRESS_ZLIB:
+			return zlib.NewReader(limitedReader)
+		case MYSQL_COMPRESS_ZSTD:
+			return zstd.NewReader(limitedReader)
+		}
+	}
+
+	return nil, nil
+}
+
+func (c *Conn) currentPacketReader() io.Reader {
```

Comment:
there're

```
	compressedReaderActive bool

	compressedReader io.Reader
```

in `Conn`. Seems we can directly check `c.compressedReader == nil` as the returned reader for `currentPacketReader`. And `compressedReaderActive` always has the same value for `c.compressedReader == nil` so we can delete it.

Reply:
When I attempted to delete `compressedReaderActive` the tests in the `client` package all began failing when I ran them with compression enabled. I think this is because `compressedReaderActive` is reset to false in `WritePacket` after writing the compressed packet. So I don't think I can delete it, or at least I feel deleting it is out of scope for this PR.

Reply:
I've updated this PR with your suggestion. It was 2am for me and I wasn't thinking clearly, but after more sleep, I realized I could easily remove the `compressedReaderActive` boolean property from `Conn`.

Reply:
take care of your health ❤️